src/Player.cpp

```diff
@@ -2264,34 +2270,64 @@ void Player::Step( int col, int row, const RageTimer &tm, bool bHeld, bool bRele
 				// bug. (It was fNoteOffset > 0.f before) -DaisuMaster
 				if( !REQUIRE_STEP_ON_HOLD_HEADS && ( fNoteOffset <= GetWindowSeconds( TW_W5 ) && GetWindowSeconds( TW_W5 ) != 0 ) )
 				{
-					score = TNS_W1;
+					// Set it to the first non-disabled window.
+					const auto &disabledWindows = m_pPlayerState->m_PlayerOptions.GetCurrent().m_twDisabledWindows;
+					if (disabledWindows.find(TW_W1) == disabledWindows.end())
+						score = TNS_W1;
+					else if (disabledWindows.find(TW_W2) == disabledWindows.end())
+						score = TNS_W2;
+					else if (disabledWindows.find(TW_W3) == disabledWindows.end())
+						score = TNS_W3;
+					else if (disabledWindows.find(TW_W4) == disabledWindows.end())
+						score = TNS_W4;
+					else if (disabledWindows.find(TW_W5) == disabledWindows.end())
+						score = TNS_W5;
+				
 					break;
 				}
 				// Fall through to default.
 			default:
 				if( (pTN->type == TapNoteType_Lift) == bRelease )
 				{
-					if(	 fSecondsFromExact <= GetWindowSeconds(TW_W1) )	score = TNS_W1;
-					else if( fSecondsFromExact <= GetWindowSeconds(TW_W2) )	score = TNS_W2;
-					else if( fSecondsFromExact <= GetWindowSeconds(TW_W3) )	score = TNS_W3;
-					else if( fSecondsFromExact <= GetWindowSeconds(TW_W4) )	score = TNS_W4;
-					else if( fSecondsFromExact <= GetWindowSeconds(TW_W5) )	score = TNS_W5;
+					const auto &disabledWindows = m_pPlayerState->m_PlayerOptions.GetCurrent().m_twDisabledWindows;
+					if(	 fSecondsFromExact <= GetWindowSeconds(TW_W1) && disabledWindows.find(TW_W1) == disabledWindows.end())	score = TNS_W1;
+					else if( fSecondsFromExact <= GetWindowSeconds(TW_W2) && disabledWindows.find(TW_W2) == disabledWindows.end())	score = TNS_W2;
+					else if( fSecondsFromExact <= GetWindowSeconds(TW_W3) && disabledWindows.find(TW_W3) == disabledWindows.end())	score = TNS_W3;
+					else if( fSecondsFromExact <= GetWindowSeconds(TW_W4) && disabledWindows.find(TW_W4) == disabledWindows.end())	score = TNS_W4;
+					else if( fSecondsFromExact <= GetWindowSeconds(TW_W5) && disabledWindows.find(TW_W5) == disabledWindows.end())	score = TNS_W5;
 				}
 				break;
 			}
 			break;
 
 		case PC_CPU:
 		case PC_AUTOPLAY:
-			score = PlayerAI::GetTapNoteScore( m_pPlayerState );
+		{
+			score = PlayerAI::GetTapNoteScore(m_pPlayerState);
+			const auto& disabledWindows = m_pPlayerState->m_PlayerOptions.GetCurrent().m_twDisabledWindows;
+			for (int i = score; i >= TNS_W5; i--)
+			{
+				TapNoteScore cur_score = (TapNoteScore)i;
+				// Downgrade the TapNoteScore if that specific window is disabled.
+				if (cur_score == TNS_W1 && disabledWindows.find(TW_W1) == disabledWindows.end())
+					score = TNS_W2;
+				else if (cur_score == TNS_W2 && disabledWindows.find(TW_W2) == disabledWindows.end())
```

Comment:
I think that should all be plain `if`s instead of `else if` to make the downgrades work correctly when multiple windows are disabled. e.g. W1 -> W2 -> W3

Reply:
Changed here and elsewhere. PTAL.